plugin/core/sessions.py

```diff
@@ -1317,6 +1399,19 @@ def m_workspace_applyEdit(self, params: Any, request_id: Any) -> None:
         self._apply_workspace_edit_async(params.get('edit', {})).then(
             lambda _: self.send_response(Response(request_id, {"applied": True})))
 
+    def m_workspace_semanticTokens_refresh(self, params: Any, request_id: Any) -> None:
+        """handles the workspace/semanticTokens/refresh request"""
+        self.send_response(Response(request_id, None))
+        debug("server requested a refresh of semantic tokens")
+        # TODO should we send a request only for the active view, or also for all other SessionBuffers of the Session?
+        active_view = sublime.active_window().active_view()
+        if not active_view:
+            return
+        session_view = self.session_view_for_view_async(active_view)
+        if not session_view:
+            return
+        session_view.session_buffer.do_semantic_tokens_async(active_view)
```

Comment:
We should refresh semantic tokens for all `SessionView`s in that session. Which I believe already happens so the TODO can be removed.

Though I don't like how it gets the active view using `active_view()`. This can get a completely unrelated view.

We should have something like this at least:

```suggestion
        for session_buffer in self._session_buffers:
            session_buffer.do_semantic_tokens_async(self.session_views_async().view)
```

Though since `sessionBuffer` is `SessionBufferProtocol`, we don't have access to `do_semantic_tokens_async()` type-wise we probably have to declare that method in `SessionBufferProtocol`.

Reply:
This suggestion doesn't work:
> AttributeError: 'generator' object has no attribute 'view'

Anyways, does the `Session.session_views_async().view` iterator only contains the primary views and is it ensured that the view always corresponds to the correct `session_buffer`? Probably it would be easier to get the view from the SessionBuffer and maybe omit the parameter and do it inside `do_semantic_tokens_async` as I had it before. I just followed the style from `_do_color_boxes_async` in the current code.

Btw, what is the purpose of this SessionBuffer / SessionBufferProtocol distinction (and the same for SessionView)? I see that SessionBufferProtocol is basically an empty interface, but for what is it used?

I also think we should refresh tokens for all SessionViews, which would happen with this change. From the docs:
> This is useful if a server detects a project wide configuration change which requires a re-calculation of all semantic tokens. Note that the client still has the freedom to delay the re-calculation of the semantic tokens if for example an editor is currently not visible.

It just happens that rust-analyzer sends the refresh request *after every change* of the document. According to https://github.com/microsoft/vscode/issues/130486 this is a workaround to enforce token updates in VSCode even for other visible (unedited) files, which may had highlighting changes due to a change in another file.
My concern is just that it is a lot of work to always `do_semantic_tokens_async()` for all files on every change, and that's why I limited it to the active file for now. Ideally the LSP specs would be so that the server could communicate in `workspace/semanticTokens/refresh` exactly the files which had token changes.

Reply:
We should probably get the number of groups with `Window.num_groups()`, iterate through all groups to get the visible views (`Window.active_view_in_group(group)`) and then send requests for the visible views (if they belong to the session). What do you think?

Reply:
> Anyways, does the `Session.session_views_async().view` iterator only contains the primary views and is it ensured that the view always corresponds to the correct `session_buffer`? Probably it would be easier to get the view from the SessionBuffer and maybe omit the parameter and do it inside `do_semantic_tokens_async` as I had it before. I just followed the style from `_do_color_boxes_async` in the current code.

There can be only one `SessionBuffer` per file (for given session).
There can be multiple `SessionView`s per file and all reference the same `SessionBuffer`.
There are multiple `SessionViews` per file in case the file has clones (`Split View`).

Color boxes is a different cases since it uses phantoms and apparently those are tied to buffer so we need to create those per `SessionBuffer`. Regions are per view so we need to create them per `SessionView`.

Omitting view parameter in `do_semantic_tokens_async` is probably OK since it should be OK to get `view.style_for_scope` from any view and we still have to update regions in all views anyway.

> Btw, what is the purpose of this SessionBuffer / SessionBufferProtocol distinction (and the same for SessionView)? I see that SessionBufferProtocol is basically an empty interface, but for what is it used?

I believe generally the idea of *Protocol interface is to expose minimal interface to places that deal with instances of those so that those users are not tempted to do shortcuts by accessing all the existing properties/methods. It should be a black box (opaque) to most places that are not supposed to mess with it.

In that case, I don't immediately see another solution so I think it would just make sense to expose `do_semantic_tokens_async` on SessionBufferProtocol.

> It just happens that rust-analyzer sends the refresh request after every change of the document. According to microsoft/vscode#130486 this is a workaround to enforce token updates in VSCode even for other visible (unedited) files, which may had highlighting changes due to a change in another file.

Not quite liking the thought of purposely breaking spec compliance to cater for specific server.
Maybe it's not that bad in practice? Even less so if server supports deltas.
Also, from reading of that issue, it doesn't look to me to be VSCode specific. Just a typical case of changes in one file possibly influencing semantic tokens in another file. Same thing would affect ST as far as I can tell.

> We should probably get the number of groups with `Window.num_groups()`, iterate through all groups to get the visible views (`Window.active_view_in_group(group)`) and then send requests for the visible views (if they belong to the session). What do you think?

Can we think of something like that if it turns to be an issue rather then possibly optimizing prematurely?

Reply:
> There can be only one SessionBuffer per file (for given session).
> There can be multiple SessionViews per file and all reference the same SessionBuffer.
> There are multiple SessionViews per file in case the file has clones (Split View).

This is clear to me, but I asked about `Session.session_views_async()`, which you used as a parameter to `session_buffer.do_semantic_tokens_async()`. I assume `Session.session_views_async()` is the collection of all SessionViews over all SessionBuffers, so passing an arbitrary view to `do_semantic_tokens_async()` of an unrelated SessionBuffer cannot work. I guess I can omit the parameter and revert it to use `SessionBuffer.some_view()` again, for the request.

---

> Not quite liking the thought of purposely breaking spec compliance to cater for specific server.

> Can we think of something like that if it turns to be an issue rather then possibly optimizing prematurely?

It's not breaking the spec, in contrary, the LSP spec explicitly mentions that the client is supposed/allowed to not send requests for all files, see https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#semanticTokens_refreshRequest:
> Note that the client still has the freedom to delay the re-calculation of the semantic tokens if for example an editor is currently not visible.

Also from the other thread it appears to me that VSCode currently does and should do the same what I suggested, described both in https://github.com/microsoft/vscode/issues/130486#issuecomment-937684422 and in https://github.com/microsoft/vscode/issues/130486#issuecomment-938590332:

> * the client will then send requests for any other ___visible file___ managed by the same provider.

(emphasis by me)

And it would feel pretty wrong to do all calculations and draw regions even for file tabs which are not visible. Let's say you have 10 files open with 100 code lines each, and there are 10 different tokens per line (rust-analyzer basically emits tokens for almost every character - excluding whitespace, but for example including commas, parentheses etc. - and it could probably be used to entirely replace the .sublime-syntax). That would mean we do 10000 function calls to decode a token and calculate the corresponding scope *for basically each keystroke*. With some effort I have added caching for this function in 0a6f7fb, but there is also more work like creating a new instance of `SemanticToken` for each token (for use in the scope name popup). If we really want to send requests for all SessionViews, then I would recommend to add a blacklist for rust-analyzer and don't allow semantic highlighting for this server until it complies with the LSP spec, or until there is a better solution in the spec to communicate the relevant files. I'd like to hear a third opinion on this @rwols

Reply:
I think this should work as expected now:
A new token request will be sent for all visible views, and for all other views which are not currently visible a flag is set and the tokens will then be updated when the view gets selected/activated.